executors/multiversx/refundExecutor.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/multiversx/mx-sdk-go/builders"
+	"github.com/multiversx/mx-sdk-go/data"
+)
+
+const (
+	getRefundTransactionsFunction    = "getRefundTransactions"
+	executeRefundTransactionFunction = "executeRefundTransaction"
+	refundTxType                     = "refund"
+)
+
+// ArgsRefundExecutor represents the DTO struct for creating a new instance of type refundExecutor
+type ArgsRefundExecutor struct {
+	TransactionExecutor    TransactionExecutor
+	ScProxyBech32Addresses []string
+	Proxy                  Proxy
+	Codec                  Codec
+	Filter                 ScCallsExecuteFilter
+	Log                    logger.Logger
+	GasToExecute           uint64
+}
+
+type refundExecutor struct {
+	*baseExecutor
+	gasToExecute uint64
+}
+
+// NewRefundExecutor creates a new instance of type refundExecutor
+func NewRefundExecutor(args ArgsRefundExecutor) (*refundExecutor, error) {
+	if args.GasToExecute < minGasToExecuteSCCalls {
+		return nil, fmt.Errorf("%w for GasToExecute: provided: %d, absolute minimum required: %d", errGasLimitIsLessThanAbsoluteMinimum, args.GasToExecute, minGasToExecuteSCCalls)
+	}
+
+	executor := &refundExecutor{
+		baseExecutor: &baseExecutor{
+			scProxyBech32Addresses: args.ScProxyBech32Addresses,
+			proxy:                  args.Proxy,
+			transactionExecutor:    args.TransactionExecutor,
+			codec:                  args.Codec,
+			filter:                 args.Filter,
+			log:                    args.Log,
+		},
+		gasToExecute: args.GasToExecute,
+	}
+
+	err := executor.checkBaseComponents()
+	if err != nil {
+		return nil, err
+	}
+
+	return executor, nil
+}
+
+// Execute will execute one step: get all pending operations, call the filter and send execution transactions
+func (executor *refundExecutor) Execute(ctx context.Context) error {
+	return executor.executeOnAllScProxyAddress(ctx, executor.executeRefundForScProxyAddress)
+}
+
+func (executor *refundExecutor) executeRefundForScProxyAddress(ctx context.Context, scProxyAddress string) error {
+	executor.log.Info("Executing refunds for the SC proxy address", "address", scProxyAddress)
+
+	pendingOperations, err := executor.getPendingRefunds(ctx, scProxyAddress)
+	if err != nil {
+		return err
+	}
+
+	filteredPendingOperations := executor.filterOperations("refundExecutor", pendingOperations)
+
+	return executor.executeRefunds(ctx, filteredPendingOperations, scProxyAddress)
+}
+
+func (executor *refundExecutor) getPendingRefunds(ctx context.Context, scProxyAddress string) (map[uint64]bridgeCore.ProxySCCompleteCallData, error) {
+	request := &data.VmValueRequest{
+		Address:  scProxyAddress,
+		FuncName: getRefundTransactionsFunction,
+	}
+
+	response, err := executor.proxy.ExecuteVMQuery(ctx, request)
+	if err != nil {
+		executor.log.Error("got error on VMQuery", "FuncName", request.FuncName,
+			"Args", request.Args, "SC address", request.Address, "Caller", request.CallerAddr, "error", err)
+		return nil, err
+	}
+	if response.Data.ReturnCode != okCodeAfterExecution {
```

Comment:
should we consider checking response and response.Data against nil here?

Reply:
added checks + removed duplicated code + added unit tests